### ISSUE TITLE
Reformatting a list expression with if/else changes the list expression

### ISSUE THREAD

**Issue by gubser** [fsprojects/fantomas#2972]
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKBtAOgOwhAlgGZgCMYAvKWAC4AWiuZiANgM6JgBMOeYAzDgC6OSABpwUdtWr5sAc1aow6CRFC8IkACTUAngAdkaSLOpjVebQDcAhswCuR5Rd4AGcRssBJbEmzUAZXwAL2QPTzAAFhcwQQsAX3C1GO09QyUTf3NPaztHJRUI0iSNSABZGwAPABlZRGqGOTpsiJJOVxi43kSLdVKdAydIBngAeUJa7EQAvWYwlK1bBydCiM4S3kgAUV9xyfmiyGZCSE6Ekr7NgfTjACNYWGYWq6X8tFXPPg3vbHYAJ2oADFZHYAHKIADuzDqzw01D%2BjjO3QuCzSQ3uj1hmkWeRWMQgkW%2BmgC%2Bhs0EQACFEIRYH9EAAFGx-GwAW0Q1EQfyxeHhiM8XTwPV4l0s13RDye31yywK%2BLAAFYiVASWTKdTafVYBDOdAbOwfFZYLqZLBsNyILzEEjBSicmKMhjJQtXniigA2JWQFXkqk0ukAVX0hj%2Buv12ENxvwpvNYEIdnY1ogQrwIuxaIdEu50reziKAHZPd61X7EABhZh61il02sXnQai0mNxthW-nnXqowYZzFSnEy95ygAchdJPvVdLKiBZt05TfjrY0AqTtv66bumd7LtlRQAnCPVb6NdXmNGlc2E23kR27WuoI6s32cx8NCR3HKvaPEABBQgcv7VlkWRsGNLUTMBk2SG8u3XHtnVxbdWhIfcxxLAIp3wOATzNM95zAiCwFTKB7Rgp07S3AcijaZDv1-Tk0JZDDHlPOVQMvG1r1XaC7w3OD%2B1zVovnfIsvz%2BWB7F8AARFh8AYv8QIRBdeCXcCVyuW9MjMTd4Io1pCXfCpKi8QgABV6GwAJaFpagAHV8HgZolQ6Njlw4tSuI0h9yP4zwSEVfSqiM0yGC2FsLKs2z7NoGM3Scxd22FTsbigUxPO07yXw9fzDOwQh8EqUZgxsBs-i2Sp9DpVhWCjbCh1ipT4pTRKhhSrS%2BOfXgSALLKACVEDgP54Aihy5UiOq8GU-DCNSdyWt4p85RIYcer62l4FBexp05cYvA5FlFCVEg8NU0V1JZexmBkaEpkBWkgOkTljK7Vr5sovd316-r4DKc7LrqG6-juuTPWgWgmTJP8AH0ITs5ojtck6ZqyZ7XTWN9DgMkTmV0UY-lqWshqipVBzGiAJuOtNEc0uaUc8TgkKyzGbGx3H8FrdbNr%2Bbbdv2ha4YSqCksgM6LvwK7EH%2BwGHqe6mENp9Z30Z5m8eob6RbFiWiqB98QbB%2BtOShmGor5xqBeapGZZ02nBPRqoADVcQpWR4FkOQCZjYnjcgzjBdmsi0vavBOD0m3KkBcT62qx3fBdt2lVGz2CKajJff6LyA4gTg-JDiTYGoABxdlSvKxBKuq2PaoTqbiO42C-bauVOEyw5wShOoqWobUGEewwpNy7B8BNbAv18ScOZ5opWLiq9%2Be98Va9T-2G66w4v2hORsDD7AI9NIJ16K%2Bw6WM2AfD8agiuqucW0rpOSNS%2Buik4JaV4uzlsHPqxNXkUeZz%2BXvZAH6q48IjnkUuNBqXs3KC3vMjWWGhOBvXRj9fA%2BxqisluPAGw5ZYBVXkC3MWl8LxT3YjPSBc9SIL3vhEPgaMIjbDKpyGSDAz7MAANKIEQPoE%2BTCfAUmZFvQmcoQHX1Nt2chLxF5FD4PTQ4FIKzYAANb7FYCJMSvhwS1kQF9JBYsi4VSqjWeSfIiEuRIQjKBPE64vSofLGRTJDx0gkqzaAfwZIgg5PAf0-dTRSWgBWZkg8CGgNJuAxOIjjDC1%2BlMXhZJ5HshmLoOYd8rGfGtrQ1WkTKTMmgLEwIswDi0OcayQw8BTjORUvDCmPtzaWJphoPgwdaFsI4QZdmP9xiyJsAopRMZXwk1iCEqu6loEW3SrwPgWdaF4LbuORA6TRZ1AAvoewZ9B66JLvomqE8FLCNnqIpJtSxlN1oUXBhbJ-B2BCgxVgAiihCLKfECQghwgmFYICKqSghHxCAA)

#### Code

```fsharp
[
    if 1 = 1 then 1 else 2
    3
]

```
            

#### Result

```fsharp
[ if 1 = 1 then
      1
  else
      2
      3 ]

```
            
#### Problem description

Fantomas reformats the list in a way that produces a different list:

```fsharp
> [
-     if 1 = 1 then 1 else 2
-     3
- ];;
val it: int list = [1; 3]

> [ if 1 = 1 then 1 else 2; 3 ];;
val it: int list = [1]
```

#### Extra information

- [x] The formatted result breaks my code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.
- [ ] I would like a release if this problem is solved.

#### Options

Fantomas main branch at 1/1/1990

Default Fantomas configuration

<sub>Did you know that you can ignore files when formatting by using a [.fantomasignore file](https://fsprojects.github.io/fantomas/docs/end-users/IgnoreFiles.html)?</sub>
<sub>PS: It's unlikely that someone else will solve your specific issue, as it's something that you have a personal stake in.</sub>

**Comment by gubser** [comment]
Hi there!

I've found a workaround by setting `fsharp_array_or_list_multiline_formatter` to `number_of_items` and have it set to 1. See [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKBtAOgOwhAlgGZgCMYAvKWAC4AWiuZiANgM6JgBMOeYAzDgC6OSABpwUdtWr5sAc1aow6CRFC8IkACTUAngAdkaSLOpjVebQDcAhswCuR5Rd4AGcRssBJbEmzUAZXwAL2QPTzAAFhcwQQsAX3C1GO09QyUTf3NPaztHJRUI0iSNSABZGwAPABlZRGqGOTpsiJJOVxi43kSLdVKdAydIBngAeUJa7EQAvWYwlK1bBydCiM4S3kgAUV9xyfmiyGZCSE6Ekr7NgfTjACNYWGYWq6X8tFXPPg3vbHYAJ2oADFZHYAHKIADuzDqzw01D%2BjjO3QuCzSQ3uj1hmkWeRWMQgkW%2BmgC%2Bhs0EQACFEIRYH9EAAFGx-GwAW0Q1EQfyxeHhiM8XTwPV4l0s13RDye31yywK%2BLAAFYiVASWTKdTafVYBDOdAbOwfFZYLqZLBsNyILzEEjBSicmKMhjJQtXniigA2JWQFXkqk0ukAVX0hj%2Buv12ENxvwpvNYEIdnY1ogQrwIuxaIdEu50reziKAHZPd61X7EABhZh61il02sXnQai0mNxthW-nnXqowYZzFSnEy95ygAchdJPvVdLKiBZt05TfjrY0AqTtv66bumd7LtlRQAnCPVb6NdXmNGlc2E23kR27WuoI6s32cx8NCR3HKvaPEABBQgcv7VlkWRsGNLUTMBk2SG8u3XHtnVxbdWhIfcxxLAIp3wOATzNM95zAiCwFTKB7Rgp07S3AcijaZDv1-Tk0JZDDHlPOVQMvG1r1XaC7w3OD%2B1zVovnfIsvz%2BWB7F8AARFh8AYv8QIRBdeCXcCVyuW9MjMTd4Io1pCXfCpKi8QgABV6GwAJaFpagAHV8HgZolQ6Njlw4tSuI0h9yP4zwSEVfSqiM0yGC2FsLKs2z7NoGM3Scxd22FTsbigUxPO07yXw9fzDOwQh8EqUZgxsBs-i2Sp9DpVhWCjbCh1ipT4pTRKhhSrS%2BOfXgSALLKACVEDgP54Aihy5UiOq8GU-DCNSdyWt4p85RIYcer62l4FBexp05cYvA5FlFCVEg8NU0V1JZexmBkaEpkBWkgOkTljK7Vr5sovd316-r4DKc7LrqG6-juuTPWgWgmTJP8AH0ITs5ojtck6ZqyZ7XTWN9DgMkTmV0UY-lqWshqipVBzGiAJuOtNEc0uaUc8TgkKyzGbGx3H8FrdbNr%2Bbbdv2ha4YSqCksgM6LvwK7EH%2BwGHqe6mENp9Z30Z5m8eob6RbFiWiqB99sA2mc-gh2BCAh-BudOZyVPhinBdmsi0vavBOEE9GqgANVxClZHgWQ5AJmNib5xqBeapGZZ02m9OdypAXE%2Btqo93xvd9pVRoDyDOOtkPbbauVOD8yOJNgagAHF2VK8rEEq6qk9q1OCKa7tSP6Lz7YgThMsOcEoTqKlqG1BhHsMKTcuwE3qq-XxJw5nmilYuKr359PxVgrOXrWLrDi-aE5GwaPsFj00gm3or7DpYzYB8PxqCK6q5xbWupuI7jl6bu2c6WjeLs5bBr6sTV5EnvWQ9ZCjxrLfC8c92ILzcoLe8yNZYaE4G9dGP18D7GqKyW48AbDllgFVeQncxZgMUuNBqadoFL0bi8V%2BRQ%2BBowiNsMqnIZIMCvswAA0ogRA%2BgL4sJ8BSZke9CZynPMQ0mpC65BwbqlbOND6aHApBWbAABrfYrARJiV8OCWsiAvooLFmXCqVVQFKlnvVeegdF5SLgWHDQfB5byKZIeOkElWbQD%2BDJEEHJ4D%2BhHqaKS0AKzMhNDVIoIj7712MMLX6Ux%2BFkiUeyGYug5jSNXp8J29DVbRMpMyaA8TAizAOPQtxrJDDwDNhAlyUCEYZypivGmtiI70I4Vwgy7M9bjAUTYZRqiYyvhJrEcRD91KwNDulXgfA870IId3cciBMmizqABfQ9gr7BIMRXIxISIimJIeYsh1SKEpPqeM9u9Cy5MLZP4OwIUGKsCEaE3C5t4gSEEOEEwrBARVSUGE%2BIQA)

**Comment by nojaf** [comment]
Hello,

It seems like the AST has changed after formatting, leading to an invalid result.
Maybe we can extend 
https://github.com/fsprojects/fantomas/blob/a49d1703900692c72ce5378c97316e2be4c01b83/src/Fantomas.Core/CodePrinter.fs#L1800-L1824

and not check for `YieldExpr` but always go multiline when there is any `Expr.IfThenElse`.

Are you interested in submitting a PR?

**Comment by gubser** [comment]
Cool! Yes, I'd like to submit one. But it's going to be next week.

**Comment by gubser** [comment]
TL;DR: I've given up setting up Fantomas dev env, sorry guys. I run into issues installing .NET 8 sdk on Ubuntu 22.04.3 and it seems complicated enough that I would easily spend the whole day trying to understand what's going on under the hood.

I'm having trouble installing .NET 8 deb packages (conflict) and using the .tar.gz packaged sdk doesn't work either.

In [Getting Started](https://fsprojects.github.io/fantomas/docs/contributors/Getting%20Started.html#Preview-SDK) to install `Release/7.0.4xx` but global.json requires at least `8.0.100-preview.7.23376.3`, so I went ahead and tried installing 8.0.

Then in [dotnet-installer](https://github.com/dotnet/installer#installers-and-binaries) it says I need to install the deps separately from [this page](https://github.com/dotnet/runtime/blob/main/docs/project/dogfooding.md#daily-builds-table) I assume. However, `sudo dpkg -i dotnet-host-x64.deb` fails with
```
dpkg: regarding dotnet-host-x64.deb containing dotnet-host:
 dotnet-host-7.0 conflicts with dotnet-host
  dotnet-host (version 8.0.0-1) is to be installed.
```
But I'm not gonna uninstall dotnet-host-7.0 because I need it for work.

I tried another approach by downloading the sdk as a .tar.gz from [this page](https://dotnet.microsoft.com/en-us/download/dotnet/8.0) and install it per instruction (extract, set DOTNET_ROOT and PATH variable).
With that approach I got further, I could do `dotnet tool restore` and `dotnet restore`. But when I ran `dotnet fsi build.fsx` it failed in CheckFormat stage on `dotnet fantomas src docs build.fsx --check` because it didn't find the fantomas tool. However, the command `fantomas src docs build.fsx --check` worked find. So I changed that from `dotnet fantomas...` to just `fantomas...`.
Now the script failed in UnitTests because it could not find my .NET 7.0.

Maybe I'll try again some later time when os-specific .NET 8 packages are out.

**Comment by nojaf** [comment]
Hi there, sorry to hear you had so many issues installing `dotnet`.
Our `Getting Started` guide should indeed refer to the latest SDK in the `global.json`.
So, you were right to go with `dotnet 8 rc 2`.
It is very sad to hear that just installing things was so painful.

It also sounds like you hit https://github.com/dotnet/sdk/issues/35989. Which is another frustrating thing for sure.

I typically follow the steps listed [here](https://learn.microsoft.com/en-us/dotnet/core/install/linux-scripted-manual#manual-install) when I need to install the SDK on Linux.

```bash
# Download the SDK
curl -O -J https://download.visualstudio.microsoft.com/download/pr/9144f37e-b370-41ee-a86f-2d2a69251652/bc1d544112ec134184a5aec7f7a1eaf9/dotnet-sdk-8.0.100-rc.2.23502.2-linux-x64.tar.gz
DOTNET_FILE=dotnet-sdk-8.0.100-rc.2.23502.2-linux-x64.tar.gz
export DOTNET_ROOT=$HOME/.dotnet 
mkdir -p "$DOTNET_ROOT" && tar zxf "$DOTNET_FILE" -C "$DOTNET_ROOT" 
export PATH=$PATH:$DOTNET_ROOT:$DOTNET_ROOT/tools
```

I'm not quite sure why the Unit tests failed. This may need another rollforward hack.

I appreciate you taking the time to look into this and feel bad for the puzzling state of dotnet.
Hopefully the situation will improve in the future.

**Comment by dawedawe** [comment]
@gubser So, you tried to mix distro packages and tar balls?
IIRC that's very likely to fail as you also experienced.
